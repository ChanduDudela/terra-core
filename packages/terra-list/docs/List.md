# Terra List

The Terra List is a structural component to vertically arrange content within list/listitems.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-list`
  - `yarn add terra-list`

## Usage

```jsx
import React from 'react';
import List, { Item } from 'terra-list';

<List isDivided={true}>
  <Item key="section-1-item-1">
    <p>blurb</p>
  </Item>
  <Item key="section-1-item-2">
    <p>blurb</p>
  </Item>
</List>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
