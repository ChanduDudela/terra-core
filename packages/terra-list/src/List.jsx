import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import styles from './List.module.scss';
import Item from './ListItem';
import Section from './ListSection';
import Subsection from './ListSubsection';
import SectionHeader from './ListSectionHeader';
import SubsectionHeader from './ListSubsectionHeader';
import Utils from './ListUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children list items passed to the component.
   */
  children: PropTypes.node,
  /**
   * Whether or not the list's child items should have a border color applied.
   */
  isDivided: PropTypes.bool,
  /**
   * Function callback for the ref of the ul.
   */
  refCallback: PropTypes.func,
  /**
   * Accessibility role of the list, defaults to 'listbox'. If creating a read-only list, pass 'none', and no role will be added.
   */
  role: PropTypes.string,
};

const defaultProps = {
  children: [],
  isDivided: false,
  role: 'listbox',
};

const List = ({
  children,
  isDivided,
  refCallback,
  role,
  ...customProps
}) => {
  const listClassNames = cx([
    'list',
    customProps.className,
  ]);

  const attrSpread = {};
  if (role && role.length > 0 && role !== 'none') {
    attrSpread.role = role;
  }
  if (isDivided) {
    attrSpread['data-list-items-divided'] = isDivided;
  }

  return (
    <ul {...customProps} {...attrSpread} className={listClassNames} ref={refCallback}>
      {children}
    </ul>
  );
};

List.propTypes = propTypes;
List.defaultProps = defaultProps;

export default List;
export {
  Item,
  Section,
  Subsection,
  SectionHeader,
  SubsectionHeader,
  Utils,
};
