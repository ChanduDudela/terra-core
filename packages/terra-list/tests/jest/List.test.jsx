import React from 'react';
import List, { Item } from '../../src/List';

// Snapshot Tests
it('should render with items', () => {
  const item1 = <Item key="123" />;
  const item2 = <Item key="124" />;
  const item3 = <Item key="125" />;
  const item4 = <Item key="126" />;
  const item5 = <Item key="127" />;
  const items = [item1, item2, item3, item4, item5];
  const shallowComponent = shallow(<List>{items}</List>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with no items', () => {
  const shallowComponent = shallow(<List />);
  expect(shallowComponent).toMatchSnapshot();
});
