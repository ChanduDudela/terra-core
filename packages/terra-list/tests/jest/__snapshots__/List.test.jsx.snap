// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with divided items 1`] = `
<ul
  className="list"
  data-list-items-divided={true}
  role="listbox"
>
  <ListItem
    isSelected={false}
    key="123"
  />
  <ListItem
    isSelected={false}
    key="124"
  />
  <ListItem
    isSelected={false}
    key="125"
  />
  <ListItem
    isSelected={false}
    key="126"
  />
  <ListItem
    isSelected={false}
    key="127"
  />
</ul>
`;

exports[`should render with items 1`] = `
<ul
  className="list"
  role="listbox"
>
  <ListItem
    isSelected={false}
    key="123"
  />
  <ListItem
    isSelected={false}
    key="124"
  />
  <ListItem
    isSelected={false}
    key="125"
  />
  <ListItem
    isSelected={false}
    key="126"
  />
  <ListItem
    isSelected={false}
    key="127"
  />
</ul>
`;

exports[`should render with no items 1`] = `
<ul
  className="list"
  role="listbox"
/>
`;
